### ISSUE TITLE
Sporadic NaN output during CPU inference

### ISSUE THREAD

**Issue by lunixbochs** [facebookresearch/flashlight#505]
I'm trying to track down an issue I'm having with flashlight. This is with a conformer model similar to the rasr conformers, and I'm using CPU inference.

For some users, maybe once per hour, multiple calls to network->forward() in a row return emissions that are the correct shape but 100% NaNs. I've done extensive debugging up to this point, and the samples going into inputFeatures are sane when this happens, and I believe the output of inputFeatures does not contain NaNs. I can also restart my process and replay the exact same input samples and I don't get NaNs on a fresh load.

My question here: is there possibly some kind of persistent state inside `forwardSequentialModuleWithPadMask` or `network->forward()` that I'm missing in my mental model? Maybe something about dnnl? Uninitialized memory? Not completely filling an array?

I'm just really not sure what would cause this to persist across a few calls to forward(), briefly, then work for a long time again. I have no reliable reproduction case, just several users who are running debug code and hit this issue regularly.

Right now I'm a little confused by `inputSizes` in `forwardSequentialModuleWithPadMask`

**Comment by lunixbochs** [comment]
Ok, I've found two causes for this.

1. PowerSpectrum is returning -inf elements sometimes. This is probably my fault.

2. `forwardSequentialModuleWithPadMask()` generates a boolean mask:
```
  auto padMask = af::iota(af::dim4(T, 1), af::dim4(1, B)) <
      af::tile(inputNotPaddedSize, T, 1);
```
Then conformer calls log() on the mask later:
```
    padMask = fl::Variable(af::log(padMaskArr), false);
```
`logf(0) == -inf`, and -inf is likely to result in the output being nan after matrix math.

I'm expecting to fix (1) myself. Can anyone explain (2) to me? It seems like a bug, I don't know why the mask would be generated as 0 or 1, then used with log() later.

**Comment by lunixbochs** [comment]
inputSize overflowed sometimes because I passed an int instead of a float

**Comment by tlikhomanenko** [comment]
FYI: attention weights are stored in log scale (so they are logits) and then you apply softmax to normalize. The way people often implement masking in attention (that some parts cannot be used for computation of attendance, like no looking into future) is setting logits of attention to -inf at positions not allowed so that after softmax you will have 0 at this position = not attend. We do exactly this in all attention places - put -inf at masking positions.
